File: clean_code_test/clean_example.js
/**
 * Clean JavaScript code that should pass linting checks
 */

/**
 * Calculate the sum of an array of numbers
 * @param {number[]} numbers - Array of numbers to sum
 * @returns {number} The sum of all numbers
 */
function calculateSum(numbers) {
  if (!Array.isArray(numbers)) {
    throw new TypeError('Input must be an array');
  }

  return numbers.reduce((acc, num) => {
    if (typeof num !== 'number') {
      throw new TypeError('All elements must be numbers');
    }
    return acc + num;
  }, 0);
}

/**
 * Class representing a data processor
 */
class DataProcessor {
  /**
   * Create a data processor
   * @param {string} name - The processor name
   * @param {Object} options - Configuration options
   */
  constructor(name, options = {}) {
    this.name = name;
    this.options = {
      batchSize: 100,
      timeout: 5000,
      ...options
    };
    this.processedCount = 0;
  }

  /**
   * Process a batch of data
   * @param {Array} data - Data to process
   * @returns {Promise<Array>} Processed data
   */
  async processBatch(data) {
    const results = [];
    
    for (const item of data) {
      const processed = await this.processItem(item);
      results.push(processed);
      this.processedCount++;
    }
    
    console.log(`Processed ${data.length} items`);
    return results;
  }

  /**
   * Process a single item
   * @param {Object} item - Item to process
   * @returns {Promise<Object>} Processed item
   */
  async processItem(item) {
    // Simulate async processing
    await new Promise(resolve => setTimeout(resolve, 10));
    
    return {
      id: item.id,
      processed: true,
      timestamp: new Date().toISOString(),
      processor: this.name,
      data: item
    };
  }

  /**
   * Get processing statistics
   * @returns {Object} Statistics object
   */
  getStatistics() {
    return {
      processorName: this.name,
      batchSize: this.options.batchSize,
      totalProcessed: this.processedCount
    };
  }
}

/**
 * Filter array by predicate function
 * @param {Array} array - Array to filter
 * @param {Function} predicate - Filter function
 * @returns {Array} Filtered array
 */
const filterArray = (array, predicate) => {
  if (!Array.isArray(array)) {
    return [];
  }
  
  if (typeof predicate !== 'function') {
    throw new TypeError('Predicate must be a function');
  }
  
  return array.filter(predicate);
};

/**
 * Utility functions for string manipulation
 */
const StringUtils = {
  /**
   * Capitalize first letter of string
   * @param {string} str - Input string
   * @returns {string} Capitalized string
   */
  capitalize(str) {
    if (typeof str !== 'string') {
      return '';
    }
    return str.charAt(0).toUpperCase() + str.slice(1);
  },

  /**
   * Check if string is empty or whitespace
   * @param {string} str - String to check
   * @returns {boolean} True if empty or whitespace
   */
  isBlank(str) {
    return !str || !str.trim();
  },

  /**
   * Truncate string to specified length
   * @param {string} str - String to truncate
   * @param {number} maxLength - Maximum length
   * @param {string} suffix - Suffix to append
   * @returns {string} Truncated string
   */
  truncate(str, maxLength, suffix = '...') {
    if (typeof str !== 'string') {
      return '';
    }
    
    if (str.length <= maxLength) {
      return str;
    }
    
    return str.slice(0, maxLength - suffix.length) + suffix;
  }
};

// Example usage
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    calculateSum,
    DataProcessor,
    filterArray,
    StringUtils
  };
}

File: clean_code_test/clean_example.py
"""Clean Python code that should pass all linting checks."""

from typing import List, Optional, Dict, Any
import logging

# Configure logging
logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)

logger = logging.getLogger(__name__)


class DataProcessor:
    """Process data with various transformations."""

    def __init__(self, name: str, batch_size: int = 100):
        """Initialize the processor.

        Args:
            name: Name of the processor.
            batch_size: Size of processing batches.
        """
        self.name = name
        self.batch_size = batch_size
        self.processed_count = 0

    def process_batch(self, data: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Process a batch of data.

        Args:
            data: List of data items to process.

        Returns:
            Processed data items.
        """
        results = []
        for item in data:
            processed = self._transform_item(item)
            results.append(processed)
            self.processed_count += 1

        logger.info(f"Processed {len(data)} items")
        return results

    def _transform_item(self, item: Dict[str, Any]) -> Dict[str, Any]:
        """Transform a single data item.

        Args:
            item: Data item to transform.

        Returns:
            Transformed data item.
        """
        return {
            "id": item.get("id", ""),
            "processed": True,
            "processor": self.name,
            "data": item,
        }

    def get_statistics(self) -> Dict[str, Any]:
        """Get processing statistics.

        Returns:
            Dictionary containing processing statistics.
        """
        return {
            "processor_name": self.name,
            "batch_size": self.batch_size,
            "total_processed": self.processed_count,
        }


def calculate_average(numbers: List[float]) -> Optional[float]:
    """Calculate the average of a list of numbers.

    Args:
        numbers: List of numbers.

    Returns:
        Average value or None if list is empty.
    """
    if not numbers:
        return None
    return sum(numbers) / len(numbers)


def filter_valid_items(
    items: List[Dict[str, Any]], required_fields: List[str]
) -> List[Dict[str, Any]]:
    """Filter items that have all required fields.

    Args:
        items: List of items to filter.
        required_fields: List of required field names.

    Returns:
        Filtered list of valid items.
    """
    valid_items = []
    for item in items:
        if all(field in item for field in required_fields):
            valid_items.append(item)

    logger.debug(f"Filtered {len(valid_items)} valid items from {len(items)}")
    return valid_items


if __name__ == "__main__":
    # Example usage
    processor = DataProcessor("example_processor")

    sample_data = [
        {"id": "1", "value": 100},
        {"id": "2", "value": 200},
        {"id": "3", "value": 300},
    ]

    processed = processor.process_batch(sample_data)
    stats = processor.get_statistics()

    print(f"Processing complete: {stats}")
    print(f"Average value: {calculate_average([100.0, 200.0, 300.0])}")
